components/script/dom/element.rs

```diff
@@ -2449,15 +2449,27 @@ impl ElementMethods for Element {
 
     /// <https://w3c.github.io/DOM-Parsing/#widl-Element-innerHTML>
     fn SetInnerHTML(&self, value: DOMString) -> ErrorResult {
-        // Step 1.
-        let frag = self.parse_fragment(value)?;
         // Step 2.
         // https://github.com/w3c/DOM-Parsing/issues/1
         let target = if let Some(template) = self.downcast::<HTMLTemplateElement>() {
             DomRoot::upcast(template.Content())
         } else {
             DomRoot::from_ref(self.upcast())
         };
+
+        // Fast path for when the value is small, doesn't contain any markup and doesn't require
+        // extra work to set innerHTML.
+        if !self.node.has_weird_parser_insertion_mode() &&
+            value.len() < 100 &&
+            !&['<', '&', '\0', '\r'].iter().any(|&c| value.contains(c))
```

Comment:
Would be nice to not go through `value` four times unconditionally.

Reply:
Do you mean iterating over value and checking if current character is in `<&\0\r` ?

Reply:
I can change it to `value.chars().any(|c| ['&', '\0', '<', '\r'].contains(&c))` if you think that's more efficient

Reply:
That probably does end up generating code that's easier to optimize.

Reply:
So I did a bit of benchmarking. What I'm proposing is faster on very small strings in release mode. However, in debug mode, perf is horrendous...

```
[eijebong@plutonium] ~   >>> rustc --test -O a.rs
[eijebong@plutonium] ~   >>> ./a --bench

running 8 tests
test current_100_length                  ... bench:          51 ns/iter (+/- 2)
test current_100_length_matching_center  ... bench:          14 ns/iter (+/- 3)
test current_10_length                   ... bench:          39 ns/iter (+/- 5)
test current_10_length_matching_center   ... bench:           7 ns/iter (+/- 0)
test proposed_100_length                 ... bench:          54 ns/iter (+/- 1)
test proposed_100_length_matching_center ... bench:          31 ns/iter (+/- 0)
test proposed_10_length                  ... bench:           5 ns/iter (+/- 0)
test proposed_10_length_matching_center  ... bench:           3 ns/iter (+/- 0)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured; 0 filtered out

[eijebong@plutonium] ~   >>> rustc --test a.rs
[eijebong@plutonium] ~   >>> ./a --bench

running 8 tests
test current_100_length                  ... bench:         267 ns/iter (+/- 17)
test current_100_length_matching_center  ... bench:         112 ns/iter (+/- 5)
test current_10_length                   ... bench:         275 ns/iter (+/- 28)
test current_10_length_matching_center   ... bench:         112 ns/iter (+/- 5)
test proposed_100_length                 ... bench:      19,971 ns/iter (+/- 1,149)
test proposed_100_length_matching_center ... bench:      10,284 ns/iter (+/- 1,097)
test proposed_10_length                  ... bench:       2,199 ns/iter (+/- 215)
test proposed_10_length_matching_center  ... bench:       1,312 ns/iter (+/- 63)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured; 0 filtered out

```

Here's the benchmark: https://gist.github.com/Eijebong/c04ed49a54c852f0e0340c5b467082a6